### ISSUE TITLE
Where raw on a joinrelation field

### ISSUE THREAD

**Issue by FredericLatour** [Vincit/objection.js#1108]
Hi,
I'm trying to apply a filter to a field in a join relation. The filter involves a MySql function, therefore I'm trying to use 'raw'.
Unfortunately it doesn't return any data (no exception is thrown)

```
HTTP/1.1 200 OK
X-Powered-By: Express
Access-Control-Allow-Origin: *
Content-Type: application/json; charset=utf-8
Content-Length: 2
ETag: W/"2-vyGp6PvFo4RvsFtPoIWeCReyIC8"
Date: Thu, 11 Oct 2018 19:32:38 GMT
Connection: keep-alive

{}
```

Here is the query

```typescript
try {
      const options = getOptionsFromUrl(req.originalUrl)
      const result = await Order.query()
        .select([
          'orders.id',
          'orders.purchase_id',
          'purchase_date',
          'orders.regname',
          'bc_company',
          'bc_email'
        ])
        .orderBy('orders.id', 'desc')
        .eager('purch_items.[lkeys, prod(idOnly).maintenance(maint)]')
        .modifyEager('purch_items', builder =>
          builder.select('id', 'purchase_id_id', 'product_id', 'product_name')
        )
        .modifyEager('purch_items.lkeys', builder =>
          builder.select(
            'purchase_item_id',
            'key',
            'maintenance_end_date',
            'license_type',
            'replace_key_id',
            'replaced_by_id'
          )
        )
        .joinRelation('purch_items.lkeys')
        //.where('purch_items:lkeys.purchase_item_id', 1006)
        .where( raw('Year(purch_items:lkeys.maintenance_end_date) = ?'), 2018)
        .where( raw('Month(purch_items:lkeys.maintenance_end_date) = ?'), 7)
        .debug()
        
      res.json(result)
    } catch (error) {
      res.json(error)
    }
```

Note that if instead of the 2 `where(raw(...`, I use `.where('purch_items:lkeys.purchase_item_id', 1006)`, 
I will retrieve the proper item (just to say that the query seems to be working apart from that).

Please note also that I tried a couple of variations to address the field: 
(eg: Year(maintenance_end_date), Year(lickeys.maintenance_end_date), ...) without any success.

Also, I don't get any debug information to the console even though no exception is thrown...

I must be doing something wrong. Anybody has a clue?

More generally, it doesn't feel always natural to deal with multiple joins/eager and filtering on 
fields that are parts of the eager/joined tables. 

I made some tests with sequelize as well (not an expert either) and those kind of queries with multiple 
eager loading and filtering are quite straightforward.

Maybe I'm not having the best approach with objections. Is there a better way than what I'm doing?

Thanks

**Comment by koskimas** [comment]
These `raw` calls should work. Using the `??` placeholder can be used for column names so that you don't need to take care of correct escaping.

```js
        .where(
          raw('Year(??) = ?', [
            "purch_items:lkeys.maintenance_end_date",
            2018
          )
        )
        .where(
          raw('Month(??) = ?', [
            "purch_items:lkeys.maintenance_end_date",
            7
          ])
        )
```

Errors are thrown just like in any javascript code. You are doing something wrong. No way to guess what that is.

As for the comments about objection being not natural or sequelize being better, what can I say? Feel free to choose the best tool for your problem.

**Comment by FredericLatour** [comment]
Thanks,

Looking at your code, I realized my mistake:
```typescript
        .where( raw('Year(purch_items:lkeys.maintenance_end_date) = ?'), 2018)
        .where( raw('Month(purch_items:lkeys.maintenance_end_date) = ?'), 7)
```
The parenthesis are not set at the right place. It should have been:

```typescript
        .where( raw('Year(purch_items:lkeys.maintenance_end_date) = ?', 2018))
        .where( raw('Month(purch_items:lkeys.maintenance_end_date) = ?', 7))
```
For the story, this does not work either but, in that case, I receive a proper ER_PARSE_ERROR message.
To make it work, I have to use your approach in order to properly escape field names.

**Things to note:**
* I was too confident that by using Typescript I would not do this kind of typo.
* I confirm that using :
```typescript
        .where( raw('Year(purch_items:lkeys.maintenance_end_date) = ?'), 2018)
        .where( raw('Month(purch_items:lkeys.maintenance_end_date) = ?'), 7)
```
will not throw any error but more surprisingly will not output any debug information to the console. I don't know if there is something to look at here but wanted to let you know.


  

> As for the comments about objection being not natural or sequelize being better, what can I say? Feel free to choose the best tool for your problem.

I am the first to admit that my english is far from perfect. However, expressing a feeling, comparing experience with another tool  and wondering if I'm having the right approach (with Objection) for those kind of queries is not the same thing as saying sequelize is better.

**What could you say?** (just some examples)
*  Can you show me how you get this query with sequelize?
OR
* yes, objection is a bit tricky in that area but shines here and there...
OR
* maybe the documentation should be somewhat improved regarding more complex queries.
OR
* Indeed you are making your life a bit difficult and you could make it simpler by doing this, that, etc ...
OR
* whatever ...

My goal is indeed to try to select the best tool for my problem (and we all know this is not always simple because each tool usually has its pro and cons). That's why I'm spending time to try the tools in a real world use case.
I'm not sure why you feld my comments was kind of a definitive judgement.

**Comment by koskimas** [comment]
Objection is what it is at this point. Its cannot be fundamentally changed anymore. Some people like it, others don't. Either use it or don't. I'm happy to listen suggestions about new features and bugfixes, but the basic structure is pretty much set in stone.

Please remember that this is open source. I using my free time to answer these questions and I don't get any money from doing this. I don't care if you use objection or not, so you are not entitled to any service here.

**Comment by FredericLatour** [comment]
Why do you even care to open source if you don't mind people using Objection or not?
This doesn't make sense to me but never mind.
Thanks anyway.

**Comment by FredericLatour** [comment]
One last thing. I don't know who this 'k-j-kleist' is but I don't see the point to come up with a thumbs up here.
Do you feel any better ? 

Either my english is really bad and I am not able to understand that my initial post was really agressive toward Objection or you guys need to relax.

Do I need to come up with my Credit Card if I ever have another question or comment?

**Comment by koskimas** [comment]
The message was not aggressive. As I originally said, there's no way I can help you if you don't like fundamental design of objection.

**Comment by koskimas** [comment]
@FredericLatour Acutally now I have to apologize. I was on mobile and didn't read the whole messages. You actually did have a question at the end of the original message. I thought you were just stating things and it seemed pointless. Sorry about this. I see a bunch of entitled stuff here and in other open source projects and I just made a snap judgement.

So what do you feel is unnatural in the query? Did you notice that you can use `joinEager` instead of `eager` + `joinRelation`?

**Comment by FredericLatour** [comment]
Thanks. That's appreciated. After re-reading my initial post in all possible ways, I just couldn't understand why the conversation went that way.
Back to objection.

I guess that my confusion comes from the fact that objection seems to offer so many ways to approach the same problem. While making my way into it, I'm always wondering if I'm taking the right path.

First thing that I was expecting for instance is that `.eager()` would allow me to specify the fields I want to retrieve. It seems more natural to me to say: "I want to load items from this relation using those fields" rather than: "I want to load items from this relation" and "modify .eager() for this relation because I don't want all fields".
In the documentation, chapter eager-loading, modifyEager is not mentionned and in the examples on the right, modifyEager is used but only for 'where' or 'order'. I suppose that if you know the knex query builder stuff, it would be obvious that one can use 'select' here. Obviously I tried to put my hand into objection without really having a look to knex.
Now, I can understand that because you have the possibility to specify fields by default at the model level and named filters as well, in most case a well thought application will mostly rely on these features rather than specifying fields at the query level. 

Another thing that bite me is that `.eager()` will not use join by default. Therefore filtering on fields that are part of the `.eager()` will not return the result you would expect from a join. 

Something that would help immensely people like me would be to have an example starting from a SQL query with 3/4 tables in relation BOTH left inner joins, with a limited set of fields retrieved for each table and some filter on fields from the joined table (one of the filter could even involve a raw approach :)). It would be assumed that data from the joined table would be displayed as a graph (eager).
Having the Objection idomatic way to make such a query would certainly clarify things up in one page.

That said, I was really not that far (a parenthesis away) from getting my query to work properly. 

You mention `joinEager` instead of `eager` + `joinRelation` but it was my understanding that `joinEager` is a left join while `joinRelation` is inner join by default. Searching further, I could read that it's not yet possible to have an 'inner `joinEager`'. Did I understand correctly?
Sounds like it could be interesting to have inner based joined for eager as well and to have some way to mix both. For instance prefixing a relation with '>' meaning we want an inner join. No prefix would default to left join:

```typescript
joinEager('>purch_items.[>lkeys, prod(idOnly).maintenance(maint)]')
```

Special mention to the "objection-graphql" module. It seem too good to be true. Most graphql helper  I have seen usually still involve writing tons of verbose code. Is that so easy? Haven't tried it yet but for that alone, objection looks like a very interesting option .


PS: After re-reading the documention I realized that I can restrict an eager relation to certain fields at the query level without using modifyEager by defining a modifier within the query.
Instead of 
```typescript
       .eager('purch_items.[lkeys, prod(idOnly).maintenance(maint)]')
        .modifyEager('purch_items', builder =>
          builder.select('id', 'purchase_id_id', 'product_id', 'product_name')
        )
```
I can write this:

```typescript
       .eager('purch_items.[lkeys(lkeysFields), prod(idOnly).maintenance(maint)]', {
          lkeysFields: builder =>
               builder.select('id', 'purchase_id_id', 'product_id', 'product_name')
       })
```
I prefer this approach even though I can imagine that modifyEager makes sense when building queries dynamically.

To sum up my experience up to now:
* the different possible approaches for doing things is a bit overwhelming initially
* once you realize that objection is built on top of Knex, things start to sound better.
* don't overlook the documentation samples at the right of the documentation. They may display some features not necessarily mentioned in the text.


**Thanks again for having re-opened the issue after realizing that I was indeed questioning my way of doing things.**

**Comment by koskimas** [comment]
It's true, objection gives a lot of freedom for people to implement things the way they want and there are often multiple ways to do one thing. The basic design principle behind objection has always been:

1. Provide helpers for common difficult relational problems
2. Always allow people to fall back to plain SQL and leverage SQL as much as possible

Objection is basically a query builder with relational helpers and not so much a traditional ORM.

This has lead to the problem you describe: there's no way to easily document all these things without making the documentation too big to read. It's already huge and doesn't cover all the use cases. Different people also try to find different things, using different words and It's not easy to make the kind of "all essential stuff in one page" -kind of tutorial you described.

> Another thing that bite me is that .eager() will not use join by default. Therefore filtering > on fields that are part of the .eager() will not return the result you would expect from a join.

Eager loading large graphs using joins is actually really inefficient. Not because of the queries, but because of the way data is transmitted from the database to the client. Try to create a three way join (three relations) with each table having 10 results and you'll see that you get 1000 result rows, each row having all the fields of all the tables. So if you do a join based eager query for one row that has three relations say `a`, `b` and `c`, and each of those relations have 10 items, the result set has 1000 rows and not the expected 31. That's alot of data to transfer. For that reason, the default is not join based.

You can actually use any join type by calling the `eagerOptions` method and setting the `joinOperation`, but that's for all joins. There's currently no way to specify different join types for each relation. I'd really like to avoid extending the relation expression language, because it's already quite complex.

Have you noticed you can define `modifiers` for model classes? Using modifiers, you can acutally do nice things like this:

```js
class BaseModel extends Model {
  static get modifiers() {
    // This example assumes you use `jsonSchema`. You can of course get
    // the list of columns anywhere else too.
    return Object.keys(this.jsonSchema.properties).reduce((modifiers, prop) => {
      modifiers[prop] = builder => builder.select(prop)
      return modifiers
    }, {})
  }
}

class Person extends BaseModel {
  static get jsonSchema() {
    return {
      type: 'object',
      properties: {
        id: { type: 'integer' },
        name: { type: 'string' }
      }
    }
  }
}

// Now you can do this in eager queries:
Person.query().eager('parents(id, name).children(id)')
```

You can of course create a modifier that selects some default set of properties and just use that always. There's also the `applyModifier` method for applying modifiers for the root query.

objection-graphql is as easy as it seems, but it has it's drawbacks when you need to do something it doesn't automatically generate. I haven't actually developed it myself in years so I don't really know what's going on there nowadays.

**Comment by koskimas** [comment]
@FredericLatour I made some improvements to the docs based on your feedback https://github.com/Vincit/objection.js/commit/2fe87572a976892ce734ea4756da4a44d328445b

**Comment by FredericLatour** [comment]
@koskimas That's a great addition. I believe it is important to mention modifyEager in the doc part. The explicit example involving select also makes it clear.

> Have you noticed you can define modifiers for model classes? Using modifiers, you can acutally do nice things like this:

Yes I had used two of them in my query (for the sake of testing). Even though it's a nice display of Objection flexibility, I don't think I would use that approach. The .eager expression could become really crowdy with each field specified at this level. Using the inline modifier (which I didn't know it was possible before yesterday) is flexible enough if I want some specific fields at the query level.

I was going to ask if there was some soft-delete feature but then I see there is some plugin.

I will now give a try try to graph insertion.

Thanks for the help.

**Comment by koskimas** [comment]
You can split eager expressions to multiple lines:

```js
Person.query().eager(`
  purch_items(id, foo, bar, baz).[
    lkeys(id, spam, eggs), 

    prod(idOnly).[
      maintenance(maint)
    ]
  ]
`)
```

**Comment by FredericLatour** [comment]
Nice astute. Still I may need 'where' and/or 'order' filter as well and I would not want to mix things/concepts (there are already so much ways to do things) and make it more difficult than necessary if I re-read my code say 1 year later. Standing behind the concept of named modifiers as it is is my preferred option. Of course that's me and one could prefer the expressiveness of specifying columns directly within the parenthesis.

I was wondering (not asking for an implementation though) : Wouldn't it be interesting to have the possibility to set eagerOptions within modifyEager like the following ?

```javascript
const people = await Person
  .query()
  .eager('[children.[pets, movies], movies]')
  .modifyEager('children.pets', builder => {
    builder.where('age', '>', 10).select('name');
   }, options: {
   joinOperation: 'innerJoin'
 }
);
```

It would make it easy to set eager joinOperation at the relation level. That said other eagerOptions do not necessarily make sense when applied to a single relation I suppose.
